RecoBTag/FeatureTools/src/SeedingTracksConverter.cc

```diff
@@ -0,0 +1,232 @@
+                tp_features_vector.clear();        
+                
+                if (reco::deltaR(it.track(), jet) > 0.4) continue;
+                
+                std::pair<bool,Measurement1D> ip = IPTools::absoluteImpactParameter3D(it, pv);
+                std::pair<bool,Measurement1D> ip2d = IPTools::absoluteTransverseImpactParameter(it, pv);
+                std::pair<double, Measurement1D> jet_dist =IPTools::jetTrackDistance(it, jetdirection, pv); 
+                TrajectoryStateOnSurface closest = IPTools::closestApproachToJet(it.impactPointState(),pv, jetdirection,it.field());
+                float length=999;
+                if (closest.isValid()) length=(closest.globalPosition() - pvp).mag();
+                
+                
+                if (!(ip.first && ip.second.value() >= 0.0 &&
+                      ip.second.significance() >= 1.0 && 
+                      ip.second.value() <= 9999. &&
+                      ip.second.significance() <= 9999. &&
+                      it.track().normalizedChi2()<5. &&
+                      std::fabs(it.track().dxy(pv.position())) < 2 &&
+                      std::fabs(it.track().dz(pv.position())) < 17 &&
+                      jet_dist.second.value()<0.07 &&
+                      length<5. )) 
+                    continue;                
+                
+                
+                btagbtvdeep::SeedingTrackInfoBuilder seedInfo;
+                seedInfo.buildSeedingTrackInfo(&(it), pv, jet, masses[selTrackCount-1], probabilityEstimator, computeProbabilities);
+
+                unsigned int neighbourTrackCount=0;
+                
+                for(auto const& tt : selectedTracks){
+                    
+                    neighbourTrackCount+=1;
+
+                    if(tt==it) continue;
+                    if(std::fabs(pv.z()-tt.track().vz())>0.1) continue;
+
+                    btagbtvdeep::TrackPairInfoBuilder trackPairInfo;
+                    trackPairInfo.buildTrackPairInfo(&(it),&(tt),pv,masses[neighbourTrackCount-1],jetdirection);
+                    sortedNeighboursMap.insert(std::make_pair(trackPairInfo.pca_distance(), trackPairInfo));
+                    
+                }
+      
+                int max_counter=0;
+
+                for(auto const& im: sortedNeighboursMap){
+                                
+                    if(max_counter>=20) break;
+                    btagbtvdeep::TrackPairFeatures tp_features;
+                    
+                    int logOffset=0;
+                    auto const& tp = im.second;
+                    
+                    tp_features.pt=(tp.track_pt()==0) ? 0: 1.0/tp.track_pt();
+                    tp_features.eta=tp.track_eta();
+                    tp_features.phi=tp.track_phi();
+                    tp_features.mass=tp.track_candMass();
+                    tp_features.dz=(logOffset+log(fabs(tp.track_dz())))*((tp.track_dz() < 0) ? -1 : (tp.track_dz() > 0));
```

Comment:
from https://github.com/cms-sw/cmssw/pull/26088#discussion_r280581764





> this and below the second part looks the same as
> `std::copysign(1.f, im.second.first.dz())`
> except for the case with dz==0, but the result in this case is not a normal number anyways:
> do you rely on the difference of inf vs nan?

Reply:
In this case (and all the signed log transformations)   ((tp.track_dz() < 0) ? -1 : (tp.track_dz() > 0)) is 0 for tp.track_dz==0.  so the final value is 0 again.
Is something like  (tp.track_dz==0)? 0: (logOffset+log(fabs(tp.track_dz())))*std::copysign(1.f, tp.track_dz())  a better solution?

Reply:
> Is something like (tp.track_dz==0)? 0: (logOffset+log(fabs(tp.track_dz())))*std::copysign(1.f, tp.track_dz()) a better solution?

I prefer this or an alternative that does not require additional repeated copy-paste of the same symbols/patterns.

Another even better option is to define
```c++
auto logWithOffset = [](float v){return (logOffset + log(std::abs(v))*std::copysign(1.f, v);};
```
and use it

Reply:
if we have to assign the lambda to some name, why not just defining an actual helper function in the class?

Reply:
> if we have to assign the lambda to some name
What do you mean?

> why not just defining an actual helper function in the class?
this will work for me as well.